crates/astria-proto/src/native/sequencer/v1alpha1.rs

```diff
@@ -3,18 +3,721 @@ use std::{
+        Ok(Self {
+            signature,
+            verification_key,
+            transaction,
+        })
+    }
+
+    #[must_use]
+    pub fn into_parts(self) -> (Signature, VerificationKey, UnsignedTransaction) {
+        let Self {
+            signature,
+            verification_key,
+            transaction,
+        } = self;
+        (signature, verification_key, transaction)
+    }
+
+    #[must_use]
+    pub fn actions(&self) -> &[Action] {
+        &self.transaction.actions
+    }
+
+    #[must_use]
+    pub fn signature(&self) -> Signature {
+        self.signature
+    }
+
+    #[must_use]
+    pub fn verification_key(&self) -> VerificationKey {
+        self.verification_key
+    }
+
+    #[must_use]
+    pub fn unsigned_transaction(&self) -> &UnsignedTransaction {
+        &self.transaction
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct UnsignedTransaction {
+    pub nonce: u32,
+    pub actions: Vec<Action>,
+}
+
+impl UnsignedTransaction {
+    #[must_use]
+    pub fn into_signed(self, signing_key: &SigningKey) -> SignedTransaction {
+        use crate::Message as _;
+        let bytes = self.to_raw().encode_to_vec();
+        let signature = signing_key.sign(&bytes);
+        let verification_key = signing_key.verification_key();
+        SignedTransaction {
+            signature,
+            verification_key,
+            transaction: self,
+        }
+    }
+
+    pub fn into_raw(self) -> raw::UnsignedTransaction {
+        let Self {
+            nonce,
+            actions,
+        } = self;
+        let actions = actions.into_iter().map(Action::into_raw).collect();
+        raw::UnsignedTransaction {
+            nonce,
+            actions,
+        }
+    }
+
+    pub fn to_raw(&self) -> raw::UnsignedTransaction {
+        let Self {
+            nonce,
+            actions,
+        } = self;
+        let actions = actions.iter().map(Action::to_raw).collect();
+        raw::UnsignedTransaction {
+            nonce: *nonce,
+            actions,
+        }
+    }
+
+    /// Attempt to convert from a raw, unchecked protobuf [`raw::UnsignedTransaction`].
+    ///
+    /// Note that actions contained in the transactions that are not set are dropped
+    /// quietly.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if one of the inner raw actions could not be converted to a native
+    /// [`Action`].
+    pub fn try_from_raw(proto: raw::UnsignedTransaction) -> Result<Self, UnsignedTransactionError> {
+        let raw::UnsignedTransaction {
+            nonce,
+            actions,
+        } = proto;
+        let n_raw_actions = actions.len();
+        let actions: Vec<_> = actions
+            .into_iter()
+            .filter_map(|raw_act| {
+                // Silently drop unset actions.
+                let converted = Action::try_from_raw(raw_act);
+                if let Err(e) = &converted {
+                    if e.is_unset() {
+                        return None;
+                    }
```

Comment:
I think this should be an error, is there any non-error case where this would happen?

Reply:
Chances are an end user directly constructs a raw protobuf with `actions: vec![Some(SequenceTx), None, None, Some(TransferTx)]` or, in go, populating the list with `nil` by accident. I felt like rejecting the entire tx when you sent a valid protobuf is too harsh if just treating "empty" actions as being absent.

Reply:
> actions: vec![Some(SequenceTx), None, None, Some(TransferTx)]

this wouldn't be a valid transaction/set of actions though, since a `None` or `nil` action doesn't make sense. also allowing `None` would result in different transactions if someone makes a `raw`, converts it to native, and converts it back to `raw` which seems bad

Reply:
Yeah, that's indeed a problem which lead me to open #310. I am going to change this to return early with an error because my change is out of scope for this PR. Thank you for raising this!